### PR TITLE
Improved Postman Compatibility

### LINKED ISSUE #13: Improved compatibility with Postman collections [bounty: $1500]
We would like to improve the workflow for users of Postman collections (v2.x) that want to load test their APIs using k6, provide an even better onboarding experience to k6 (and in extension Load Impact), by extending the postman-to-k6 converter to take the following Postman collection features into account in the conversion process. A completed solution will also include tests and docs (updated README).

**Related issues**:
- https://github.com/loadimpact/postman-to-k6/issues/7: Convert collection that uses Environment Variables
- https://github.com/loadimpact/postman-to-k6/issues/12: Add support for variables in body

## Cookies
Setting cookies properly if they’re explicitly defined in the collection, see [docs](https://www.getpostman.com/docs/v6/postman/sending_api_requests/cookies).

**Relevant links**:
- https://docs.k6.io/docs/cookies
- https://www.getpostman.com/docs/v6/postman/sending_api_requests/cookies

## TLS Client Certificates
Setting up use of client certificates in k6 if defined in the collection, see [docs](https://www.getpostman.com/docs/v6/postman/sending_api_requests/certificates).

**Relevant links**:
- https://docs.k6.io/docs/ssl-tls-client-certificates
- https://www.getpostman.com/docs/v6/postman/sending_api_requests/certificates

## Authentication
Support the most common authentication options supported by Postman, see [docs](https://www.getpostman.com/docs/v6/postman/sending_api_requests/authorization).

The authentications we want to support:

- No auth (already supported in the converter)
- Bearer token
- HTTP Basic (already supported in the converter)
- HTTP Digest
- OAuth 1 / 2
    - See: TODO KB example script
- AWS Signature v4
    - See https://support.loadimpact.com/4.0/test-scripting/examples/#aws-signature-v4-authentication (will be more performant once we have support for binary output for the hashing algorithms in the crypto package: https://github.com/loadimpact/k6/issues/856)
- NTLM Authentication

This should also include the case when the authentication is set to “Inherit auth from parent”.

**Relevant links**:
- https://github.com/loadimpact/postman-to-k6/tree/master/lib/postman/auth
- https://www.getpostman.com/docs/v6/postman/sending_api_requests/authorization

## Environment/global variables
Support environment and global variables defined in the collection, see [docs](https://www.getpostman.com/docs/v6/postman/environments_and_globals/variables). The conversion to k6 JS must make sure to adhere to the scoping rules defined in https://learning.getpostman.com/docs/postman/environments_and_globals/variables/#variable-scopes.

In general variables in the request URL, params, headers and body will have the form {{someVariable}}. These needs to be set by properly resolving the variable.

**Relevant links**:
- https://docs.k6.io/docs/environment-variables
- https://www.getpostman.com/docs/v6/postman/environments_and_globals/variables

### Dynamic variables
The conversion needs to support Postman’s dynamic variables that can be present in the request URL, params, headers and body:
- `{{$guid}}`: Adds a v4 style GUID
- `{{$timestamp}}`: Adds the current timestamp
- `{{$randomInt}}`: Adds a random integer between 0 and 1000

### Data variables
The conversion needs to support what’s called “data variables” in Postman, see [docs](https://learning.getpostman.com/docs/postman/collection_runs/working_with_data_files/). This is parameterization data loaded from a CSV or JSON file and then used in request URL, params, headers, body or pre-/post-request scripts.

### In scripts
In scripts you get variables (global, environment, dynamic or data) by using one of the APIs, see Scripting below.

## Scripting
Postman has support for JS scripting in the form of pre-request scripts and post-response test scripts. By providing a shim layer in the converted script to support the Postman Sandbox execution environment, where pre-/post-requests scripts are executed, we should be able to provide a much more useful conversion tool from Postman to k6 for doing load testing.

### Shimming
The shim layer could either be embedded in the converted script (or as a module that the converter spits out and the main file loads) or fetched from k6 github perhaps as a remote import, something like:

```js
import postman from “github.com/loadimpact/k6/master/shims/postman-sandbox.js”;
```

### Pre-/Post-request scripts/tests
These are currently appended as comments in the k6 script, but will now need to be inserted into k6 verbatim, with the necessary Postman APIs made available by the shim layer. See [pre-request scripts docs](https://www.getpostman.com/docs/v6/postman/scripts/pre_request_scripts) and [post-request test scripts](https://www.getpostman.com/docs/v6/postman/scripts/test_scripts) for more info.

Support both pre-request scripts specified for a collection, a folder and an individual request (respecting the [execution order of scripts](https://learning.getpostman.com/docs/postman/scripts/intro_to_scripts/#execution-order-of-scripts)).

### Postman Sandbox APIs
The following Postman Sandbox APIs needs to be supported/shimmed:

- It will include making some third-party libs available by default (see list in [API docs](https://www.getpostman.com/docs/v6/postman/scripts/postman_sandbox_api_reference)). This could probably be handled by inserting them as imports to be loaded by the [supported CDN/Github loaders in k6](https://github.com/loadimpact/k6/tree/master/loader).
    - [Lodash library](https://lodash.com/)
    - [cheerio library](https://cheerio.js.org/)
    - [CryptoJS library](https://code.google.com/p/crypto-js/)
- **General**
    - `postman.getResponseHeader(headerName)`
    - `xml2Json(xmlString)`
- **Environment and global variables**
    - `postman.setEnvironmentVariable(variableName, variableValue)`
    - `postman.getEnvironmentVariable(variableName)`
    - `postman.setGlobalVariable(variableName, variableValue)`
    - `postman.getGlobalVariable(variableName)`
    - `postman.clearEnvironmentVariable(variableName)`
    - `postman.clearGlobalVariable(variableName)`
    - `postman.clearEnvironmentVariables()`
    - `environment` (a global object variable used by the functions above)
    - `globals` (a global object variable used by the functions above)
- **Dynamic variables**
    - `{% raw %}{{$guid}}{% endraw %}`: Adds a v4 style GUID
    - `{% raw %}{{$timestamp}}{% endraw %}`: Adds the current timestamp.
    - `{% raw %}{{$randomInt}}{% endraw %}`: Adds a random integer between 0 and 1000
- **Cookies**
    - `responseCookies` (array variable with response cookies for current response)
    - `postman.getResponseCookie(cookieName)`
- **Request/response related properties**
    - `request` (object with request attributes)
        - `data` (key/value form data object)
        - `headers` (key/value object with request headers)
        - `method`
        - `url`
    - `responseHeaders` (key/value object with response headers)
    - `responseBody` (string, use JSON.parse or xml2json to parse it)
    - `responseTime` (number, total response time in milliseconds)
    - `responseCode` (object)
        - `code`
        - `name`
        - `detail`
    - `tests` (object with the users “checks”/”asserts”, boolean results of logical tests)
    - `iteration` (number, the current test run index)
- **Data files**
    - `data` (object with parsed CSV or JSON data, each VU and iteration in k6 would need to fetch a new row of data, see [this blog post](http://blog.getpostman.com/2014/10/28/using-csv-and-json-files-in-the-postman-collection-runner/))
- All APIs in the Postman Sandbox API reference (the `pm` object, which is only available in Postman, not Newman): https://learning.getpostman.com/docs/postman/scripts/postman_sandbox_api_reference/
    - Except the `require()`’ing of Nodejs modules, that should result in an error message saying “Can’t load module X, Node.js modules aren’t supported in k6”
    - `pm.sendRequest(url|requstObject, callback)` this call would need to be synchronous in k6, so it wouldn’t return until the callback function is done executing.

### PR DESCRIPTION
**Pull request by mul1sh**
Fixes #13  by modifying the `converter` and its `headers` to better support postman.

### CODE REVIEW

File: lib/postman/auth/awsv4.js [hunk 1]
```diff
@@ -1,7 +1,17 @@
...
+};
 
 module.exports = {
-  enabled: false,
-  header: function (options) {
+  enabled: true,
+  signedHeaders: function ({service, path}) {
+    if (service && path) {
+    	const { service, path}
+    	return aws4.sign({ service, path }, AWS_CREDS)
```
Comment (robingustafsson) [inline lib/postman/auth/awsv4.js]: The AWS v4 `Authorization` header needs to be calculated at test script runtime, not at point of conversion/code-generation. So you'd need to generate code that imports and uses a third-party lib for handling AWS v4 auth, see https://support.loadimpact.com/4.0/test-scripting/examples/#aws-signature-v4-authentication.

This issue from a sister project is also relevant here: https://github.com/loadimpact/jmeter-to-k6/issues/3

Comment (mul1sh) [reply lib/postman/auth/awsv4.js]: ok got it, thanks for the feedback

**Comment by mul1sh** [comment]
@robingustafsson let me close this, because i've seen another branch with more changes than I have locally, so it makes no sense for me to push to this PR anymore, thanks 🙂

**Comment by robingustafsson** [comment]
@mul1sh Makes sense, thanks for giving it a shot 👏

**Pull request closed without merging**